SoftwareCo/SoftwareCo/Managers/DocEventManager.cs

```diff
@@ -52,112 +50,146 @@ private async void InitPluginDataIfNotExists()
+        public async void LineChangedAsync(TextPoint start, TextPoint end, int hint)
         {
-            // wrapper for a file path
-            FileInfo fi = new FileInfo(docPath);
-            String fileName = fi.FullName;
-            if (!IsTrueEventFile(fileName))
+            // only allow single line or bulk paste. visual sudio marks up the document
+            // when opening the file so we need to prevent these change events from coming through
+            bool isSameLinePaste = start.Line == end.Line && start.LineCharOffset < start.LineLength && end.LineCharOffset > start.LineCharOffset;
+            bool isBulkPaste = start.AtStartOfLine && end.AtEndOfLine && !end.AtStartOfLine && start.Line < end.Line;
+            if (isSameLinePaste || isBulkPaste)
             {
-                return;
+                SoftwareCoPackage package = PackageManager.GetAsyncPackage();
+                if (package != null)
+                {
+                    await package.JoinableTaskFactory.SwitchToMainThreadAsync();
+                }
+
+                if (doc == null)
+                {
+                    doc = (start.DTE != null && start.DTE.ActiveDocument != null) ? start.DTE.ActiveDocument : await PackageManager.GetActiveDocument();
+                }
+
+                if (doc == null)
+                {
+                    return;
+                }
+
+                string fileName = doc.FullName;
+                if (!IsTrueEventFile(fileName))
+                {
+                    return;
+                }
+
+                InitPluginDataIfNotExists();
+                _pluginData.InitFileInfoIfNotExists(fileName);
+
+                PluginDataFileInfo pdfileInfo = _pluginData.GetFileInfo(fileName);
+
+                if (pdfileInfo == null)
+                {
+                    return;
+                }
+
+                UpdateFileInfoMetrics(pdfileInfo, start, end, null, null);
             }
-            InitPluginDataIfNotExists();
-            _pluginData.InitFileInfoIfNotExists(fileName);
         }
 
-        public async void AfterKeyPressedAsync(
-            string Keypress, TextSelection Selection, bool InStatementCompletion)
+        public async void BeforeKeyPressAsync(string Keypress, TextSelection Selection, bool InStatementCompletion, bool CancelKeypress)
         {
-            String fileName = await PackageManager.GetActiveDocumentFileName();
+            SoftwareCoPackage package = PackageManager.GetAsyncPackage();
+            if (package != null)
+            {
+                await package.JoinableTaskFactory.SwitchToMainThreadAsync();
+            }
+
+            if (doc == null)
+            {
+                doc = (Selection.DTE != null && Selection.DTE.ActiveDocument != null) ? Selection.DTE.ActiveDocument : await PackageManager.GetActiveDocument();
+            }
+
+            if (doc == null)
+            {
+                return;
```

Comment:
Should this line move to 128?

Reply:
i'll add comments but there's a method called "OnChangeAsync" that should set the local "doc" when tabs change, but if its not set yet by the time it hits one of these other listener handlers we'll try to set it. if its still not set we'll fall out of the method. its a bit of space and conditions but the alternative is exceptions that can crash the IDE